### DISCUSSION TITLE
Client-side image transforms (or, CustomJSTransform for non-scalar columns)

### DISCUSSION THREAD

**Discussion by notestaff** [bokeh/bokeh#13224] [General]
Hello all,

I’d like to show an Image to which an image transform is applied client-side before rendering. The parameters of the transform are controlled by UI widgets. When writing a CustomJSTransform for this, I ran into the following issue: how do I correctly allocate a “blank” array to hold the results? The xs variable in this case holds an array of one Float64Array2d, so the output should be the same type. Deep-copying xs using JSON.parse(JSON.stringify(xs)) did not work (and was very slow).   Would it be hard to add a copy() method for the objects BokehJS uses to represent 2D images?

Thanks!

**Comment by bryevdv** [comment]
cc @mattpap for guidance on this question

**Comment by mattpap** [comment]
JS  has no concept of deep cloning. Using `JSON` module doesn't preserve types and is slow. BokehJS has `core/util/cloneable` module, which provides deep cloning functionality. However, it currently doesn't support `ndarrays` fully (i.e. `ndarrays` don't support `clonable` protocol). I would start an issue for this.
